Validate cluster-name provided by file

L3n41c opened the pull request:
### What does this PR do?

Validate the cluster-name provided by file so that the kubelet hostname provider built by concatenating the node name with the cluster name doesn’t result in a non-RFC1123 name.

The rule enforced by this change is exactly the same as the one enforced by GKE.

### Motivation

If somebody tries to put an underscore in the cluster_name, the hostname built by the kubelet hostname provider is not RFC1123 valid and, as such, is rejected as a potential hostname.
If, in addition, the agent is running inside a kubernetes POD which hasn’t access to the IaaS metadata server, the agent will default to using the hostname seen from inside its POD, which happens to be the POD name.

Using the POD name as agent hostname is an issue because is changes at each redeployment of the agent.

L3n41c committed:
Validate cluster-name provided by file

so that the kubelet hostname provider built by concatenating the node name with the cluster name doesn’t result in a non-RFC1123 name.
The rule enforced by this change is exactly the same as the one enforced by GKE.

L3n41c commented:
cc @hkaj

hkaj commented on pkg/util/kubernetes/clustername/clustername.go:
```diff
@@ -15,6 +17,10 @@ import (
 )
 
+var (
+	validClusterName = regexp.MustCompile(`^[a-z]([a-z0-9\-]{0,38}[a-z0-9])?$`)
```
Could you add a comment there to say that this matches the naming rule in GKE? This way if the rule there changes, other contributors will understand that this line should be updated as well

L3n41c replied:
Documentation added.
With the link to GKE documentation as well as with a more human-friendly description of what the regular expression is supposed to match.

cohenyair reviewed (commented):
Do we want to document this requirement ("Name must start with a lowercase letter followed by up to 39 lowercase letters, numbers, or hyphens, and cannot end with a hyphen") somewhere in our public docs? This might be a good place I think: https://github.com/DataDog/datadog-agent/blob/d23862f86b39ba2abdb4a7a9c56d50ed4640d120/pkg/config/config_template.yaml#L1054

hkaj commented:
ah yes that's a good point, let's update the doc (cc @l0k0ms ) and the config template as well 👍

L3n41c committed:
Document the restriction on the cluster name

L3n41c committed:
Fix clustername unit tests

l0k0ms reviewed (dismissed):
Approved doc wise 👍

L3n41c committed:
Add test

ogaca-dd reviewed (commented):
LGTM

L3n41c committed:
Fix config_template: cluster_name

The `cluster_name` parameter has been introduced in commit
https://github.com/DataDog/datadog-agent/commit/4fd0cc08a0e005536073b2c0e6d9d5ff2dafead3
The parameter is named `cluster_name` everywhere in the code but it was written
`clustername` in the config_template.

L3n41c committed:
Fix `clustername` --> `cluster_name` typo
